### ISSUE TITLE
Discuss dependency injection

### ISSUE THREAD

**Issue by samreid** [phetsims/tasks#952]
In https://github.com/phetsims/wave-interference/issues/79#issuecomment-416045828 we were considering creating a composite type DoubleHeadedArrowWithBarsNode.

A type definition like https://github.com/phetsims/wave-interference/blob/f116b8b2d3f1aa7731d3866fe5d796916734e9bc/js/common/view/DoubleHeadedArrowWithBarsNode.js is inflexible. It is hard-coded to use Line and ArrowNode, and has a clumsy way to pass options through to those type. A dependency-injection approach would be much more flexible (where the nodes are passed in instead of created by DoubleHeadedArrowWithBarsNode) provides clients with full generality. However, a dependency-injection approach reveals that the logic provided by DoubleHeadedArrowWithBarsNode is trivial: space two nodes and put another between them. I don't think we need to factor out an implementation for this, and we should leave it to our design teams to encourage consistent design for this kind of component.

For example, see the change in ScrollingChartNode.js in the https://github.com/phetsims/wave-interference/commit/00fb894e5ef9e6fd5160b70b61bedc1ed1b70ee7. It is clearer how the options pass through to the Line and ArrowNode types, and only adds a 3 lines of code to get full generality and clarity (and we are able to eliminate 66 lines of DoubleHeadedArrowWithBarsNode.js).

Another way of phrasing this issue: we have had many discussions about how to properly pass through options to subcomponents. The idea proposed here is to pass through the subcomponent itself instead of passing through options and hardcoding a type to be instantiated.

Since this logic is applicable elsewhere, I'm inclined to discuss this at developer meeting. In particular, it will be important to decide where to draw the line: when should we use this pattern?

Since writing this issue originally, I also discovered `Checkbox.createTextCheckbox` is using the dependency-creation pattern (to coin a term opposite to dependency-injection pattern).  In this case is it warranted to avoid having to say `new Text()` in the 17 call sites?

**Comment by pixelzoom** [comment]
Dependency injection is one tool in the toolbox. Understanding it (and it's many forms) and knowing when to use it is what's important.  So maybe that's what this discussion is about.  

Re `Checkbox.createTextCheckbox`, "dependency creation" is a stretch; I would call that a convenience method.  I believe that I created that method, and I've since avoided using it.  I think code is clearer when the client creates the `Text` explicitly.

**Comment by samreid** [comment]
> Re Checkbox.createTextCheckbox, "dependency creation" is a stretch;

I'm coining a new phrase here.  "Dependency creation" means the function creates the concrete subtype.  "Dependency injection" means the function takes components as arguments.  `Checkbox.createTextCheckbox` takes `{string} text` and `{object} textOptions` and has this line of code:

```js
var textNode = new Text( text, textOptions );
```

Hence I have classified that as "Dependency creation".  If this was using "Dependency Injection", instead of passing in `text, textOptions` arguments, the client would pass in a `Text` instance (or indeed any Node).

**Comment by pixelzoom** [comment]
Yes, I understand that "dependency creation" is a term that you coined.  I'm just saying that I don't think a new term is necessary, and possibly confuses things.

**Comment by jonathanolson** [comment]
Sounds like something that can be helpful. We wouldn't want to overuse it, because there are many cases where only one type is used, or more detailed information is needed. Passing in factory methods (and using options instead of required parameters) may be more general.

**Comment by samreid** [comment]
We discussed this and decided that some places dependency injection is appropriate, and some places it is not.  We will have to decide on a case by case basis where to use this pattern in common code.  We considered changing Slider.js to take a SliderTrack argument, and this simplifies the options pass-through, but ultimately makes a more complex client-side API, and leaks implementation details to the client about SliderTrack.js.  This conversation was good to help bring this dimension up for discussion, so developers will be aware of it.  Before making major changes to common code to convert to dependency injection, it should be discussed with the team.  Closing for now.
